core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java

```diff
@@ -353,7 +356,28 @@ public static boolean isTablesEnumeratingTable(InformationSchemaTable table)
                     .flatMap(prefix -> tables.get().stream()
                             .filter(this::isLowerCase)
                             .map(table -> new QualifiedObjectName(catalogName, prefix.getSchemaName().get(), table)))
-                    .filter(objectName -> !isColumnsEnumeratingTable(informationSchemaTable) || metadata.getTableHandle(session, objectName).isPresent() || metadata.getView(session, objectName).isPresent())
```

Comment:
Could `getTableHandle` be redirection aware?
In this case it seems that we don't really care about intermediate `QualifiedObjectName` and it makes code more complicated.

Generally, if redirection is like a symlink, then in OS you can still refer to it using original name and don't really notice difference until file listing (detailed).

Reply:
> Could getTableHandle be redirection aware?

hmm, do you mean change getTableHandle signature to add a boolean or always redirect within `Metadata#getTableHandle`? I think we'd want to atleast avoid the latter to keep DDLs as is.

Reply:
Why would we need a boolean within `Metadata#getTableHandle`? For a feature flag? Could that check be inside `io.trino.metadata.MetadataManager#getTableHandle`?

Reply:
Current scope is to only redirect for scans and columnlisting. The signature change in `Metadata#getTableHandle` would be to indicate whether the caller wants to look for redirected tablehandle or the original one. (e.g. in DDLs)

Reply:
What about adding method like `getTableHandleForScan`?

Reply:
yeah, that seems reasonable. so basically we replace combination of `Metadata#redirectTable` + `Metadata#getTableHandle` with a new single method, and still keep `ConnectorMetadata#redirectTable` as is. let me see how that looks. 

Any thoughts on `ConnectorMetadata#redirectTable` ?  Should we name it differently to avoid confusion with `applyTableScanRedirect`? something like `linkToTable`, although I don't have good suggestions.

Reply:
@sopel39 well, we still need a way to know the redirected table name in the analyzer. e.g. look at `StatementAnalyzer#visitTable`, where access control related code should concern the target.

```
                name = metadata.redirectTable(session, name);
                analysis.addEmptyColumnReferencesForTable(accessControl, session.getIdentity(), name);
                tableHandle = metadata.getTableHandle(session, name);
```

So just returning a `TableHandle` without the name won't be enough.

Reply:
I think the general scheme should be that:
1. we would perform redirection as late as possible within analyzer. E.g don't redirect object name if not needed, but use `getRedirectedTableHandle` instead.
2. Try to refer using original object name within analyze if possible.

This way:
1. We won't confuse users with different names than in query
2. We should be able to minimize scope of changes.

Thus I think we need
1. `getRedirectedTableHandle` and use them in places where we don't really need to know redirected object name.
2. `getRedirectedObjectName and use them in places where we really need to know redirected object name, e.g: `io.trino.sql.analyzer.StatementAnalyzer.Visitor#analyzeFiltersAndMasks`

Reply:
@sopel39 applied your suggestion.